dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewScope.kt

```diff
@@ -91,10 +97,18 @@ internal open class RumViewScope(
         }
     }
 
-    private var lastMemoryVitalInfo: VitalInfo? = null
+    private var lastMemoryInfo: VitalInfo? = null
     private var memoryVitalListener: VitalListener = object : VitalListener {
         override fun onVitalUpdate(info: VitalInfo) {
-            lastMemoryVitalInfo = info
+            lastMemoryInfo = info
+        }
+    }
+
+    private var refreshRateScale: Double = 1.0
+    private var lastFrameRateInfo: VitalInfo? = null
+    private var frameRateVitalListenr: VitalListener = object : VitalListener {
+        override fun onVitalUpdate(info: VitalInfo) {
+            lastFrameRateInfo = info
```

Comment:
As I understand from [`VitalMonitor`](https://github.com/DataDog/dd-sdk-android/blob/master/dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/vitals/VitalMonitor.kt), the listener (`onVitalUpdate`) is called on the same thread as producer (`onNewSample`), so how do we ensure thread safety for `lastFrameRateInfo`?

As far as I can see, it gets written from `VitalFrameCallback's` `doFrame(frameTimeNanos: Long)` (so Choreographer's thread) and we read it here in `RumViewScope` (so from RUM scopes' thread).

Reply:
the `VitalInfo` is a data class and therefore immutable. There is no synchronisation there so basically whenever the `onVitalUpdate` is called it will make a copy of it so there is no synchronisation needed. If in the meantime the `lastFrameRateInfo` will be changed in the `VitalMonitor` from the `Choreographer` instantiation thread it will not affect this copy it will just change the value of the reference in the `VitalMonitor`

Reply:
I'm not concerned about `VitalInfo's` internal state, but about the `lastFrameRateInfo` property itself. Are properties atomic in Kotlin by default 💭? 

At least in Swift this will crash due to race condition: we can't simultaneously read a property on one thread while writing to it on another. We need to implement atomicity for such properties to ensure both reads and writes are exclusive and avoid partial reads.

Reply:
they're not but when that method will be called `onVitalUpdate(info: VitalInfo)` the `info` parameter will be passed as a copy from the `VitalMonitor` and IMO it should not cause any problems.

Reply:
I think I understand what you are referring to now...it can happen actually that this value to be changed from one stack entry to another and it can actually alter the values in the `sendView` method. @xgouchet maybe we could actually use a final copy there when computing the `refreshRateAverage` and `refreshRateMin`

Reply:
And also I think we will need to keep this in an `AtomicReference` otherwise you could have issues when one thread is trying to read and the other is writing.

Reply:
Rephrasing my question - given that property in Kotlin:
```kotlin
private var foo: SomeDataClass? = null
```
is it safe to read and write that property from two threads? This is what bothers me, as this is not thread safe in Swift (it will crash either on `self.foo` write or read due to partial memory access - as properties are not atomic by default). I don't see any code addressing this 💭.

Reply:
Yes, a reference is always atomic, so it will never crash. The only possible side effect is one thread reading a value that's "obsolete" because the reference was changed in another thread;